docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,691 @@
+       ...
+       for _, msg := range tx.GetMsgs() {
+           if sdk.MsgTypeURL(msg) != types.URLMsgPayForData {
+               continue
+           }
+
+           pfd, ok := msg.(*types.MsgPayForData)
+           if !ok {
+               app.Logger().Error("Msg type does not match MsgPayForData URL")
+               continue
+           }
+
+           if err = pfd.ValidateBasic(); err != nil {
+               ...
+               return abci.ResponseProcessProposal{
+                   Result: abci.ResponseProcessProposal_REJECT,
+               }
+           }
+
+           commitment, err := inclusion.GetCommit(cacher, dah, int(malleatedTx.ShareIndex), shares.MsgSharesUsed(int(pfd.MessageSize)))
+           if err != nil {
+               ...
+               return abci.ResponseProcessProposal{
+                   Result: abci.ResponseProcessProposal_REJECT,
+               }
+           }
+
+           if !bytes.Equal(pfd.MessageShareCommitment, commitment) {
+               ...
+               return abci.ResponseProcessProposal{
+                   Result: abci.ResponseProcessProposal_REJECT,
+               }
+           }
+       }
+   }
+   ...
+   return abci.ResponseProcessProposal{
+       Result: abci.ResponseProcessProposal_ACCEPT,
+   }
+}
+```
+
+Lastly, we also need to check that each valid `MsgPayForData` has a corresponding message, and that there are no unexpected messages.
+
+- If a `MsgPayForData` is added to the square, then its corresponding message must also be included.
+- There must not be a message without a `MsgPayForData` (does this need to be a rule? cc @adlerjohn).
+
+```go
+func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+   ...
+   // iterate over all of the MsgPayForData transactions and ensure that they
+   // commitments are subtree roots of the data root.
+   commitmentCounter := 0
+   for _, rawTx := range req.BlockData.Txs {
+       // iterate through the transactions and check if they are malleated
+       ...
+       for _, msg := range tx.GetMsgs() {
+           ...
+           commitmentCounter++
+       }
+   }
+
+   // compare the number of PFDs and messages, if they aren't
+   // identical, then  we already know this block is invalid
+   if commitmentCounter != len(req.BlockData.Messages.MessagesList) {
+       ...
+       return abci.ResponseProcessProposal{
+           Result: abci.ResponseProcessProposal_REJECT,
+       }
+   }
+   ...
+}
+```
+
+## Future Improvements
+
+The current implementation performs many different estimation and calculation steps. It might be possible to amortize these calculations to each transaction, which would make it a lot easier to confidently arrange an optimal block.
+
+## Status
+
+{Deprecated|Proposed|Accepted|Declined}
+
+## Consequences
+
+### Positive
+
+- Trust minimized message inclusion checks
+- Ability to create message inclusion fraud proofs
```

Comment:
```suggestion
- Ability to create message inclusion fraud proofs
- Ability to sign the right commitments to messages without interacting with the block producer
```

Reply:
> Ability to sign the right commitments to messages without interacting with the block producer

I didn't list this originally because this is currently what is happening. The PFD signers are still signing over correct commitments in that the validators can check for message inclusion using them, just not in a trust minimized way

Reply:
Makes sense, thanks a lot